src/v/archival/adjacent_segment_merger.cc

```diff
@@ -56,83 +56,116 @@ adjacent_segment_merger::run(retry_chain_node& rtc, run_quota_t quota) {
-        model::offset so = _last;
-        if (so == model::offset{} && _is_local) {
-            // Local lookup, start from local start offset
-            so = local_start_offset;
-        } else {
-            // Remote lookup, start from start offset in the manifest (or 0)
-            so = _archiver.manifest().get_start_offset().value_or(
-              model::offset{0});
+    for (int i = 0; i < max_reuploads_per_run; i++) {
+        if (!_enabled || _as.abort_requested()) {
+            co_return result;
         }
-        vlog(
-          _ctxlog.debug,
-          "Searching for adjacent segment run, start: {}, is_local: {}, "
-          "local_start_offset: {}",
-          so,
-          _is_local,
-          local_start_offset);
-        for (auto it = manifest.segment_containing(so); it != manifest.end();
-             it++) {
-            if (
-              !_is_local && it->second.committed_offset >= local_start_offset) {
-                // We're looking for the remote segment
+        if (result.remaining <= 0) {
+            co_return result;
+        }
+        ss::gate::holder h(_gate);
+        vlog(_ctxlog.debug, "Adjacent segment merger run begin");
+        auto scanner = [this](
+                         model::offset local_start_offset,
+                         const cloud_storage::partition_manifest& manifest)
+          -> std::optional<adjacent_segment_run> {
+            adjacent_segment_run run(_archiver.get_ntp());
+            auto [low_watermark, high_watermark] = get_low_high_segment_size(
+              _archiver.get_local_segment_size(),
+              _min_segment_size,
+              _target_segment_size);
+            model::offset so = _last;
+            if (so == model::offset{} && _is_local) {
+                // Local lookup, start from local start offset
+                so = local_start_offset;
+            } else {
+                // Remote lookup, start from start offset in the manifest (or 0)
+                so = _archiver.manifest().get_start_offset().value_or(
+                  model::offset{0});
+            }
+            vlog(
+              _ctxlog.debug,
+              "Searching for adjacent segment run, start: {}, is_local: {}, "
+              "local_start_offset: {}, low watermark: {}, high watermark: {}",
+              so,
+              _is_local,
+              local_start_offset,
+              low_watermark,
+              high_watermark);
+            for (auto it = manifest.segment_containing(so);
+                 it != manifest.end();
+                 it++) {
+                if (
+                  !_is_local
+                  && it->second.committed_offset >= local_start_offset) {
+                    // We're looking for the remote segment
+                    break;
+                }
+                auto [key, meta] = *it;
+                if (!run.maybe_add_segment(meta, high_watermark)) {
+                    continue;
+                }
```

Comment:
nit: maybe just

```
if (run.maybe_add_segment(...)) {
    break;
}
```

Also maybe worth adding a comment indicating that returning `true` means that this run is ready to be uploaded. It's easy to misinterpret this name as having the semantics "Returns true if a segment was added" instead of "Returns true if the run is uploadable".

Alternatively, maybe it's worth separating a method to add segments from the check for whether we can upload based on max_size.